### PR TITLE
4882 r50a add visa resident without phn

### PR DESCRIPTION
**Pull request by daveb-hni**
This contains the development for R50a screen which includes:

- Add student resident without PHN screen which has the following sections:
    Person search
    Select person
    Enter details
    Confirmation

- Integration with Add student with PHN when a PHN is found on previous person search.
- API endpoints for search and enroll
- API Web Config update to allow adding a trusted certificate to the SSL context

### CODE REVIEW

File: frontend/tests/e2e/roles/roles.js [hunk 1]
```diff
@@ -5,6 +5,6 @@ export const regularAccUser = Role(SITE_UNDER_TEST, async t => {
     await t
         .click('#zocial-moh_idp')
         .typeText('#username', 'hnweb1')
-        .typeText('#password', process.env.HNWEB1_TESTCAFE_PASSWORD)
+        .typeText('#password', 'hello123')
```
Comment (tschia) [inline frontend/tests/e2e/roles/roles.js:8]: please revert this change

**Review by weskubo-cgi — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java [hunk 2]
```diff
@@ -63,6 +56,9 @@
 	@Value("${hcim.cert.password}")
 	private String hcimCertPassword;
 	
+	@Value("classpath:${hcim.trust.file}")
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java:59]: Moved "classpath:" prefix for HCIM files to the properties file

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java [hunk 3]
```diff
@@ -159,21 +137,56 @@ public WebClient rapidWebClient() throws HNWebException {
...
 			        .keyManager(keyManagerFactory)
 			        .build();
 		} catch (IOException | CertificateException | UnrecoverableKeyException | KeyStoreException | NoSuchAlgorithmException e) {
 	        logger.error("Error creating SSL Context for {} due to {}.", url, e.getMessage());
 	        throw new HNWebException(ExceptionType.SSL_FAILURE, e);
 		}
 	}
+
+	private KeyManagerFactory createKeyManagerFactory(Resource certFile, String certPassword)
+			throws KeyStoreException, IOException, NoSuchAlgorithmException, CertificateException, UnrecoverableKeyException {
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java:164]: Sounds cleaner, did same for both called methods

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java [hunk 4]
```diff
@@ -159,21 +137,56 @@ public WebClient rapidWebClient() throws HNWebException {
...
+
+	private TrustManagerFactory createTrustManagerFactory(Resource trustedCertFile) throws CertificateException,
+			FileNotFoundException, IOException, KeyStoreException, NoSuchAlgorithmException {
+		
+		CertificateFactory certificateFactory = CertificateFactory.getInstance(CERTIFICATE_TYPE_X509);
+		X509Certificate hcimCert = (X509Certificate)certificateFactory.generateCertificate(new FileInputStream(trustedCertFile.getFile()));
+		
+		KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+		keyStore.load(null); // The KeyStore instance does not need to come from a file.
+		keyStore.setCertificateEntry(HCIM_CA_CERT, hcimCert);
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java:183]: Yes, it should be fully generic. Added parameter for trusted cert alias and also renamed HCIM specific variable.

File: frontend/src/components/coverage/enrollment/NameSearch.vue [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+          <AppInput :e-model="v$.firstName" id="firstName" label="First Name" type="text" v-model.trim="firstName" />
+        </AppCol>
+      </AppRow>
+      <AppRow>
+        <AppCol class="col3">
+          <AppInput :e-model="v$.secondName" id="secondName" label="Second Name (Optional)" type="text" v-model.trim="secondName" />
+        </AppCol>
+      </AppRow>
+      <AppRow>
+        <AppCol class="col4">
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearch.vue:20]: Made all fields the same size.

File: frontend/src/components/coverage/enrollment/NameSearch.vue [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      surname: '',
+      firstName: '',
+      secondName: '',
+      dateOfBirth: null,
+      gender: '',
+      searching: false,
+    }
+  },
+  computed: {
+    inputClass() {
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearch.vue:65]: Removed unused property

File: frontend/src/components/coverage/enrollment/NameSearch.vue [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+      return {
+        'error-input': this.v$.gender.$error,
+      }
+    },
+  },
+  methods: {
+    async submitForm() {
+      this.searching = true
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearch.vue:73]: Added missing code to clear errors. The other code, search and handle result, is in the function called on emit in parent component.

File: frontend/src/components/coverage/enrollment/NameSearch.vue [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      try {
+        const isValid = await this.v$.$validate()
+        if (!isValid) {
+          this.$store.commit('alert/setErrorAlert')
+          this.searching = false
+          return
+        }
+        console.log(`Person [${this.firstName}]`)
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearch.vue:81]: Will remove

File: frontend/src/components/coverage/enrollment/NameSearch.vue [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        this.$emit('search-for-candidates', {
+          surname: this.surname,
+          givenName: this.firstName,
+          secondName: this.secondName,
+          dateOfBirth: dayjs(this.dateOfBirth).format(API_DATE_FORMAT),
+          gender: this.gender,
+        })
+      } catch (err) {
+        console.log(`Error ${err}`)
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearch.vue:90]: Will remove

File: frontend/src/components/coverage/enrollment/NameSearchCandidate.vue [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+<script>
+export default {
+  name: 'NameSearchCandidate',
+  props: {
+    candidate: Object,
+  },
+  created() {
+    this.CURRENT_RECORD = 'CURRENT RECORD'
+    this.PREVIOUS_RECORD = 'PREVIOUS RECORD'
+  },
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearchCandidate.vue:27]: Agreed, overkill declaring separately.

File: frontend/src/components/coverage/enrollment/NameSearchCandidate.vue [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  data() {
+    return {
+      submitting: false,
+    }
+  },
+  computed: {
+    formatDetailsLine1() {
+      let details = ''
+      if (this.candidate) {
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearchCandidate.vue:36]: This component should always have a Candidate, made it required and removed check.

File: frontend/src/components/coverage/enrollment/NameSearchCandidate.vue [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+          return this.CURRENT_RECORD
+        case 'C':
+          return this.PREVIOUS_RECORD
+        default:
+          return ''
+      }
+    },
+    formatDetailsLine2() {
+      let details = ''
+      if (this.candidate) {
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearchCandidate.vue:53]: Fixed same as part of above change

File: frontend/src/components/coverage/enrollment/NameSearchCandidate.vue [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        if (this.candidate.identifierTypeCode) {
+          details = details + ' ' + this.candidate.identifierTypeCode
+        }
+      }
+      return details
+    },
+    formatDetailsLine3() {
+      let address = ''
+      if (this.candidate) {
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearchCandidate.vue:74]: Fixed same as part of above change

File: frontend/src/components/coverage/enrollment/NameSearchCandidate.vue [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        name = name + ', ' + this.candidate.givenName
+      }
+      if (this.candidate.secondName) {
+        name = name + ' ' + this.candidate.secondName
+      }
+      return name
+    },
+  },
+  methods: {
+    selectCandidate() {
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearchCandidate.vue:111]: It was using an emit but it did require the two levels which is cumbersome and difficult code to follow so changed to have logic here as it can access store and navigate directly from here.

File: frontend/src/components/coverage/enrollment/NameSearchCandidate.vue [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+      this.submitting = true
+      console.log(`selected candidate [${this.candidate.surname}] [${this.candidate.phn}] [${this.candidate.address1}]`)
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/NameSearchCandidate.vue:113]: Will remove

File: frontend/src/components/coverage/enrollment/ResidentDetails.vue [hunk 16]
```diff
@@ -193,54 +185,36 @@ export default {
...
-        { text: 'U.S.A', value: 'US' },
-        { text: 'Yukon', value: 'YT' },
-        { text: 'British Columbia', value: 'BC' },
-      ]
-    },
+    this.priorResidenceOptions = PRIOR_RESIDENCES
+
+    //populate data on component load
+    this.telephone = this.resident.telephone
+    this.address1 = this.resident.address1
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/coverage/enrollment/ResidentDetails.vue:198]: Using a resident object led to js rendering errors. Updated to use store getters and only populate fields that data exists in store for.

File: frontend/src/components/template/TheNavBar.vue [hunk 17]
```diff
@@ -42,15 +42,20 @@
...
-  data: function() {
+  data: function () {
     return {
       eligibilityDropdownKey: 0,
       coverageEnrollmentDropDownKey: 0,
     }
   },
   methods: {
+    resetCoverageEnrollment: function () {
+      console.log('resetting Coverage Enrollment')
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/template/TheNavBar.vue:53]: Removed

File: frontend/src/components/ui/AppRadioButtonGroup.vue [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+import AppRadioButton from './AppRadioButton.vue'
+export default {
+  name: 'AppRadioButtonGroup',
+  components: {
+    AppRadioButton,
+  },
+  computed: {
+    buttonClass() {
+      return {
+        'error-input': this.eModel.$error,
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/ui/AppRadioButtonGroup.vue:22]: looks good, thanks

File: frontend/src/views/coverage/enrollment/AddVisaResidentWithPHN.vue [hunk 19]
```diff
@@ -41,19 +41,20 @@ export default {
...
-    ;(this.PAGE_ACTION = {
+    this.PAGE_ACTION = {
       PHN_SEARCH: 'PHN_SEARCH',
-      STUDENT_REGISTRATION: 'STUDENT_REGISTRATION',
+      REGISTRATION: 'REGISTRATION',
       CONFIRMATION: 'CONFIRMATION',
-    }),
-      (this.pageAction = this.PAGE_ACTION.PHN_SEARCH)
+    }
+    console.log(`route pageAction ${this.$route.query.pageAction}`)
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/views/coverage/enrollment/AddVisaResidentWithPHN.vue:49]: Will be removed

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java [hunk 20]
```diff
@@ -159,24 +131,57 @@ public WebClient rapidWebClient() throws HNWebException {
...
+	private SslContext getSSLContext(String url, Resource certFile, String certPassword, Resource trustFile) throws HNWebException {
 
-			KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KEY_MANAGER_FACTORY_TYPE_SUN_X509);
-			keyManagerFactory.init(keyStore, certPassword.toCharArray());
+		try {
+			KeyManagerFactory keyManagerFactory = createKeyManagerFactory(certFile, certPassword);			
 			
+			TrustManagerFactory trustManagerFactory = null;
+			if (trustFile != null) {
+				trustManagerFactory = createTrustManagerFactory(trustFile, HCIM_CA_CERT);
```
Comment (weskubo-cgi) [inline backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java:144]: For this to be truly re-usable I think the trustedCertAlias should be a method parameter as well.

**Review by weskubo-cgi — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java (hunk 20 above)
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/config/WebClientConfig.java:144]: Added parameter for trust alias to make it re-usable

**Review by weskubo-cgi — APPROVED** [review]
